### PR TITLE
Adding network monitor capabilities for prerouting and postrouting ebrules

### PR DESCRIPTION
**Pull request by pjohnst5**
**What this PR does / why we need it**:
This PR adds monitoring of prerouting and postrouting rules in ebtables. 
- It will delete any unwanted rules added to prerouting and postrouting (i.e. if the rule was not added through a pod). 
- It will add back any rules that were incorrectly deleted from prerouting and postrouting (i.e. if the rule was not deleted through a pod).

**Testing**
- Added rogue rule via terminal, saw network monitor wait one iteration before deleting rogue rule
- Deleted rule in rogue fashion via terminal, saw network monitor wait one iteration before adding rule back from state
- Flushed rules via terminal in rogue fashion, saw network monitor wait one iteration before adding all rules back
- Made new pods, network monitor didn't touch those rules, left them as was and saved these new pod rules in state.
- Deleted these new pod rules in rogue fashion via terminal, network monitor added them back after one iteration
- Deleted pods normally, network monitor didn't touch missing rules, did not try to add them back
- Tried adding old rule that had been added by  pod in the past, network monitor removed it
- Did nothing for 5 minutes. Network monitor did nothing to rules.
- Tested wget pod to pod connectivity, both within same node and different nodes
- Tested wget pod connectivity to google.com

Ebtable functionality was also cleaned up by taking common functionality of set methods and placing it in one runEbCmd method.

**Which issue this PR fixes** 
An adaptation of [this PR](https://github.com/Azure/azure-container-networking/pull/159) from two years ago:

### CODE REVIEW

File: cni/azure-linux.conflist [hunk 1]
```diff
@@ -9,6 +9,9 @@
          "ipsToRouteViaHost":["169.254.20.10"],
          "ipam":{
             "type":"azure-vnet-ipam"
+         },
```
Comment (tamilmani1989) [inline cni/azure-linux.conflist]: i dont think its required

File: cni/netconfig.go [hunk 2]
```diff
@@ -67,6 +67,11 @@ type NetworkConfig struct {
 		Address       string `json:"ipAddress,omitempty"`
 		QueryInterval string `json:"queryInterval,omitempty"`
 	}
+	NetworkMonitor struct {
```
Comment (tamilmani1989) [inline cni/netconfig.go]: not needed. network monitor runs as separate daemonset..should not be part of cni

File: cni/network/network.go [hunk 3]
```diff
@@ -263,6 +263,10 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 
 	log.Printf("[cni-net] Read network configuration %+v.", nwCfg)
 
+	if err := startMonitorIfNotRunning(nwCfg); err != nil {
```
Comment (tamilmani1989) [inline cni/network/network.go]: remove this..we dont want to start network monitor from cni

File: cni/network/network_linux.go [hunk 4]
```diff
@@ -85,6 +89,43 @@ func addSnatInterface(nwCfg *cni.NetworkConfig, result *cniTypesCurr.Result) {
 	}
 }
 
+func startMonitorIfNotRunning(nwCfg *cni.NetworkConfig) error {
```
Comment (tamilmani1989) [inline cni/network/network_linux.go]: remove this

File: cni/network/network_windows.go [hunk 5]
```diff
@@ -118,6 +118,9 @@ func setEndpointOptions(cnsNwConfig *cns.GetNetworkContainerResponse, epInfo *ne
 func addSnatInterface(nwCfg *cni.NetworkConfig, result *cniTypesCurr.Result) {
 }
 
+func startMonitorIfNotRunning(nwCfg *cni.NetworkConfig) error {
```
Comment (tamilmani1989) [inline cni/network/network_windows.go]: remove

File: cnms/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,6 @@
+FROM ubuntu:latest
+RUN apt -y update
+RUN apt install -y ebtables
```
Comment (tamilmani1989) [inline cnms/Dockerfile]: Also add RUN apt-get -y upgrade after apt -y update

File: cnms/cnmspackage/monitor2rules_linux.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	currentStateRulesMap map[string]string) error {
+
+	for rule := range networkMonitor.AddRulesToBeValidated {
+		if _, ok := currentStateRulesMap[rule]; !ok {
+			delete(networkMonitor.AddRulesToBeValidated, rule)
+		}
+	}
+
+	networkMonitor.addRulesNotExistInMap(currentStateRulesMap, currentEbtableRulesMap)
+
```
Comment (tamilmani1989) [inline cnms/cnmspackage/monitor2rules_linux.go]: no space

File: cnms/cnmspackage/monitor2rules_linux.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	currentStateRulesMap map[string]string) error {
+
+	for rule := range networkMonitor.DeleteRulesToBeValidated {
+		if _, ok := currentEbtableRulesMap[rule]; !ok {
+			delete(networkMonitor.DeleteRulesToBeValidated, rule)
+		}
+	}
+
+	networkMonitor.deleteRulesNotExistInMap(currentEbtableRulesMap, currentStateRulesMap)
+
```
Comment (tamilmani1989) [inline cnms/cnmspackage/monitor2rules_linux.go]: no space

File: cnms/cnmspackage/monitor2rules_linux.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
+package cnms
+
+import (
+	"github.com/Azure/azure-container-networking/ebtables"
+	"github.com/Azure/azure-container-networking/log"
+)
+
```
Comment (tamilmani1989) [inline cnms/cnmspackage/monitor2rules_linux.go]: can you add an comment on top of each function on what it does

File: cnms/cnmspackage/monitor2rules_linux.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+		}
+	}
+
+	networkMonitor.deleteRulesNotExistInMap(currentEbtableRulesMap, currentStateRulesMap)
+
+	return nil
+}
+
+func generateL2RulesMap(currentEbtableRulesMap map[string]string, chainName string) error {
+
```
Comment (tamilmani1989) [inline cnms/cnmspackage/monitor2rules_linux.go]: no space

File: cnms/cnmspackage/monitor2rules_linux.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	table := ebtables.Nat
+	rules, err := ebtables.GetEbtableRules(table, chainName)
+
```
Comment (tamilmani1989) [inline cnms/cnmspackage/monitor2rules_linux.go]: no space

File: cnms/service/networkmonitor.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+func main() {
+	// Initialize and parse command line arguments.
+	acn.ParseArgs(&args, printVersion)
+
+	logLevel := acn.GetArg(acn.OptLogLevel).(int)
+	logTarget := acn.GetArg(acn.OptLogTarget).(int)
+	logDirectory := acn.GetArg(acn.OptLogLocation).(string)
+	timeout := acn.GetArg(acn.OptIntervalTime).(int)
+	vers := acn.GetArg(acn.OptVersion).(bool)
+
```
Comment (tamilmani1989) [inline cnms/service/networkmonitor.go]: no space

File: ebtables/ebtables.go [hunk 13]
```diff
@@ -5,61 +5,98 @@ package ebtables
...
 
-	"github.com/Azure/azure-container-networking/log"
 	"github.com/Azure/azure-container-networking/platform"
 )
 
 const (
 	// Ebtables actions.
 	Append = "-A"
 	Delete = "-D"
+	// Ebtables
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: // Ebtables table

File: ebtables/ebtables.go [hunk 14]
```diff
@@ -5,61 +5,98 @@ package ebtables
...
+	chain := PreRouting
+	rule := fmt.Sprintf("-p IPv4 -i %s --ip-dst %s -j dnat --to-dst %s --dnat-target ACCEPT",
+		interfaceName, ipAddress.String(), macAddress.String())
+
+	return runEbCmd(table, action, chain, rule)
+}
+
+// SetEbRule sets any given eb rule
+func SetEbRule(table, action, chain, rule string) error {
+
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: no space

File: network/monitor_linux.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+	cnms "github.com/Azure/azure-container-networking/cnms/cnmspackage"
+	"github.com/Azure/azure-container-networking/ebtables"
+	"github.com/Azure/azure-container-networking/log"
+)
+
+func (nm *networkManager) monitorNetworkState(networkMonitor *cnms.NetworkMonitor) error {
+
+	currentEbtableRulesMap, err := cnms.GetEbTableRulesInMap()
+
```
Comment (tamilmani1989) [inline network/monitor_linux.go]: no space

File: network/monitor_linux.go [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+	if err != nil {
+		log.Printf("GetEbTableRulesInMap failed with error %v", err)
+		return err
+	}
+
+	currentStateRulesMap := nm.AddStateRulesToMap()
+
```
Comment (tamilmani1989) [inline network/monitor_linux.go]: no space

File: network/monitor_linux.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+	networkMonitor.CreateRequiredL2Rules(currentEbtableRulesMap, currentStateRulesMap)
+	networkMonitor.RemoveInvalidL2Rules(currentEbtableRulesMap, currentStateRulesMap)
+
+	return nil
+}
+
+func (nm *networkManager) AddStateRulesToMap() map[string]string {
+
```
Comment (tamilmani1989) [inline network/monitor_linux.go]: no space

File: networkmonitor/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,6 @@
+FROM ubuntu:latest
```
Comment (tamilmani1989) [inline networkmonitor/Dockerfile]: i dont think this file is needed..we already have under cnms/service

File: networkmonitor/main.go [hunk 19]
```diff
@@ -0,0 +1,153 @@
+// Copyright 2017 Microsoft. All rights reserved.
```
Comment (tamilmani1989) [inline networkmonitor/main.go]: same here..not needed..we aleady have cnms/service

File: cni/azure-linux.conflist (hunk 1 above)
Comment (pjohnst5) [reply cni/azure-linux.conflist]: I think it is, when this comma wasn't there, `network_linux.go` threw an error saying  "NetworkMonitor is not set in conflist"
```
netmon := nwCfg.NetworkMonitor
		if netmon.Name == "" {
			return fmt.Errorf("NetworkMonitor is not set in conflist")
		}
```

File: cni/network/network_linux.go [hunk 20]
```diff
@@ -18,6 +18,7 @@ import (
 const (
 	snatInterface  = "eth1"
 	infraInterface = "eth2"
+	binPath        = "/opt/cni/bin"
```
Comment (tamilmani1989) [inline cni/network/network_linux.go]: is this needed?

File: cnms/cnmspackage/monitor2rules_linux.go [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+func (networkMonitor *NetworkMonitor) CreateRequiredL2Rules(
+	currentEbtableRulesMap map[string]string,
+	currentStateRulesMap map[string]string) error {
+
+	for rule := range networkMonitor.AddRulesToBeValidated {
+		if _, ok := currentStateRulesMap[rule]; !ok {
+			delete(networkMonitor.AddRulesToBeValidated, rule)
+		}
+	}
+	networkMonitor.addRulesNotExistInMap(currentStateRulesMap, currentEbtableRulesMap)
```
Comment (tamilmani1989) [inline cnms/cnmspackage/monitor2rules_linux.go]: space after for loop..there should always be blank after for and if

File: ebtables/ebtables.go [hunk 22]
```diff
@@ -5,61 +5,97 @@ package ebtables
...
 
-	"github.com/Azure/azure-container-networking/log"
 	"github.com/Azure/azure-container-networking/platform"
 )
 
 const (
 	// Ebtables actions.
 	Append = "-A"
 	Delete = "-D"
+	// Ebtables tables.
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: // Ebtable tables. also update above.not your change

File: ebtables/ebtables.go [hunk 23]
```diff
@@ -5,61 +5,97 @@ package ebtables
...
+
+// SetVepaMode sets the VEPA mode for a bridge and its ports.
+func SetVepaMode(bridgeName string, downstreamIfNamePrefix string, upstreamMacAddress string, action string) error {
+	table := Nat
+	chain := PreRouting
+
+	if !strings.HasPrefix(bridgeName, downstreamIfNamePrefix) {
+		rule := fmt.Sprintf("-i %s -j dnat --to-dst %s --dnat-target ACCEPT", bridgeName, upstreamMacAddress)
+
+		err := runEbCmd(table, action, chain, rule)
```
Comment (tamilmani1989) [inline ebtables/ebtables.go]: if err := runEbCmd(table, action, chain, rule); err != nil {
    return err
}

File: network/monitor_linux.go [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+// monitorNetworkState compares current ebtable nat rules with state rules and matches state.
+func (nm *networkManager) monitorNetworkState(networkMonitor *cnms.NetworkMonitor) error {
+	currentEbtableRulesMap, err := cnms.GetEbTableRulesInMap()
+	if err != nil {
+		log.Printf("GetEbTableRulesInMap failed with error %v", err)
+		return err
+	}
+	currentStateRulesMap := nm.AddStateRulesToMap()
```
Comment (tamilmani1989) [inline network/monitor_linux.go]: space after if

File: cni/network/network_linux.go (hunk 20 above)
Comment (pjohnst5) [reply cni/network/network_linux.go]: Not really, I can take it out

**Comment by tamilmani1989** [comment]
windows tests are flaky..no changes in windows made..so merging this PR

**Pull request merged**